tests/helpers_spec.lua

```diff
@@ -0,0 +1,161 @@
+        helpers.process_config(lang_config, nil, { some_parameter = "some-value" }, "nix_pkg_name").some_parameter
+      )
+    end)
+
+    it("merges default_config values when user config override is present", function()
+      assert.same(
+        "some-value",
+        helpers.process_config(
+          lang_config, { cmd = { "ls_user_cmd" } }, { some_parameter = "some-value" }, "nix_pkg_name"
+        ).some_parameter
+      )
+    end)
+
+    it("on_new_config does not error when lang config has no on_new_config function", function()
+      local config = helpers.process_config(lang_config, nil, empty_default_config, "nix_pkg_name")
+      config.on_new_config(config, "/some/root/path")
+      assert.same(
+        { "nix-shell", "-p", "nix_pkg_name", "--run", "'ls_original_cmd'" },
+        config.cmd
+      )
+    end)
+
+    it("on_new_config keeps original cmd if call back does not update it", function()
+      local lang_config_with_on_new_config = {
+        document_config = {
+          default_config = {
+            cmd = { "ls_original_cmd" },
+            on_new_config = function(new_config, root_path)
+            end
+          }
+        }
+      }
+      local config = helpers.process_config(lang_config_with_on_new_config, nil, empty_default_config, "nix_pkg_name")
+      config.on_new_config(config, "/some/root/path")
+      assert.same({ "nix-shell", "-p", "nix_pkg_name", "--run", "'ls_original_cmd'" }, config.cmd)
+    end)
+
+    it("on_new_config updates cmd when callback appends to it (e.g. omnisharp)", function()
+      local lang_config_with_on_new_config = {
+        document_config = {
+          default_config = {
+            cmd = { "ls_original_cmd" },
+            on_new_config = function(new_config, root_path)
+              table.insert(new_config.cmd, "--path")
+              table.insert(new_config.cmd, root_path)
+            end
+          }
+        }
+      }
+      local config = helpers.process_config(lang_config_with_on_new_config, nil, empty_default_config, "nix_pkg_name")
+      config.on_new_config(config, "/some/root/path")
+      assert.same(
+        { "nix-shell", "-p", "nix_pkg_name", "--run", "'ls_original_cmd' '--path' '/some/root/path'" },
+        config.cmd
+      )
+    end)
+
+    it("on_new_config updates cmd when callback build a new one (e.g. eslint, codeqlls)", function()
+      local lang_config_with_on_new_config = {
+        document_config = {
+          default_config = {
+            cmd = { "ls_original_cmd" },
+            on_new_config = function(new_config, root_path)
+              new_config.cmd = { "new_cmd", "--path", root_path }
+            end
+          }
+        }
+      }
+      local config = helpers.process_config(lang_config_with_on_new_config, nil, empty_default_config, "nix_pkg_name")
+      config.on_new_config(config, "/some/root/path")
+      -- Encoding current behavior, the presence of `ls_original_cmd` seems wrong, might need to fix it
```

Comment:
@zoriya Added some tests and discovered potentially incorrect behavior. I think the difference is that omnisharp pushes to the existing cmd, but configs for other servers can build a completely new cmd.

Reply:
Yes, it might be a better idea to call this function before modifing the command for nix-shell.

Reply:
If you are interested in contributing again feel free to open a PR. I think it should be much easier to refactor things now with tests.